### ISSUE TITLE
Crash in evaluator on `hacknix` repo

### ISSUE THREAD

**Issue by roberth** [hercules-ci/hercules-ci-agent#356]
**Description**

Evaluator crashed.

https://hercules-ci.com/github/hackworthltd/hacknix/jobs/30

```
FatalError {fatalErrorMessage = "Worker failed with exit status: -11"}
```

**To Reproduce**

See https://github.com/hercules-ci/hercules-ci-agent/issues/231#issuecomment-999969758

I will try to reproduce this by running it myself; hopefully reproducing the crash so I can debug it.

**Logs**

**Platform / Version**

See https://github.com/hercules-ci/hercules-ci-agent/issues/231#issuecomment-999969758

Best to go to https://hercules-ci.com/dashboard and click on the agents' tab for the account you're interested in.
**hercules-ci-agent --help** version

**Comment by dhess** [comment]
Right. This happened on the initial run, and if I recall, nearly every (or in fact every?) derivation failed. Then I re-evaluated the job and all of the `x86_64-linux` jobs successfully built, leaving only the macOS jobs that are still failing.

When I added this repo and ran the first job, I hadn't yet given the Hercules CI agent permission to read our private S3 binary cache, because I didn't realize that Hercules CI bypassed the Nix daemon to fetch derivations directly from the caches. I've since corrected that, but is it possible that's why the initial evaluation failed?

**Comment by dhess** [comment]
It appears possibly not. I just pushed a change to the problematic branch and the evaluator reported `-11` again.

**Comment by roberth** [comment]
I've reproduced the crash. It happens during (memory) garbage collection. I will investigate further.
Potential resolutions, one of:
 - Fix a memory bug in Nix
 - Fix a memory bug in this repo (agent or the "cnix" bindings)
 - Downgrade to Nix 2.3
 - Upgrade to Nix >=2.5.1 or master
 - Split evaluation of into multiple processes. Besides other benefits, this reduces the reliance on memory GC, which isn't always super effective anyway

I will continue to investigate these.

Possible workaround: disable GC. Worker processes exit after each eval job, so memory does not leak indefinitely. On NixOS, add:
```
systemd.services.hercules-ci-agent.environment.GC_DONT_GC = "1";
```

Evaluation tasks do not run on darwin, so you don't have to change those agents for this workaround.

**Comment by roberth** [comment]
I've found the problem and here's a workaround for until [the fix in nix](https://github.com/NixOS/nix/pull/5875) is merged and released.

```nix
    systemd.services.hercules-ci-agent = {
      serviceConfig.LimitSTACK = 40 * 1024 * 1024;
    };
```

Without the fix, the largest derivation attribute can only be half the remaining stack size, or about 20MB in the example above. After the fix, it's unlimited. You could remove the stack size increase at that point, but it's pretty harmless, thanks to virtual memory.

Evidence from darwin suggests that `alloca`-style stack allocation overflows can skip the stack protection page and cause corruption in other memory, explaining why the earlier crash happened during GC.
Perhaps more importantly, hacknix evaluation just succeeded for me locally, also suggesting that these crashes had the same cause.

**Comment by dhess** [comment]
Well I'm glad that, if nothing else, maybe you've finally got a reproducible case for https://github.com/NixOS/nix/issues/3605!
